keps/sig-network/2438-dual-stack-apiserver/README.md

```diff
@@ -0,0 +1,593 @@
+### User Stories
+
+#### Story 1 - External Clients
+
+As a cluster administrator, I want my apiserver to be reachable by
+both IPv4-only and IPv6-only clients _outside the cluster_, without
+them needing to use a proxy or NAT mechanism.
+
+#### Story 2 - APIServer Tooling
+
+As a Kubernetes add-on developer or Kubernetes service provider, I
+want to be able to reliably determine the IP addresses being used by
+the apiserver in a user-created cluster, so that I can correctly
+create and configure related infrastructure. eg:
+
+  - generating a TLS certificate that includes all of the IP addresses
+    that the apiserver is listening on as `subjectAltName`s
+
+  - pointing an external load balancer to the apiserver endpoints
+
+  - pointing an external DNS name to the apiserver endpoints
+
+#### Story 3 - Wrong-Single-Stack Internal Clients
+
+As a developer, I want to run a "wrong-single-stack pod" in a
+dual-stack cluster (that is, a pod that is single-stack of the
+cluster's secondary IP family), and have it be able to connect to the
+apiserver.
+
+This is kind of an odd case, but [it has been seen in the wild], in
+the case of a dual-stack IPv6-primary cluster that contains some nodes
+that are on an IPv4-only network, which can therefore only host
+IPv4-only pods.
+
+[it has been seen in the wild]: https://github.com/kubernetes/enhancements/issues/563#issuecomment-814560270
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+In theory, an administrator might be running an apiserver on a
+dual-stack host but not want it to be exposed as a dual-stack service.
+However, like most golang-based servers, the apiserver _listens_ on
+both IP families by default unless explicitly configured to only
+listen on a single IP, so the administrator in this case would already
+be losing, and just not realizing it. Making the apiserver advertise
+its secondary IP address in this case would arguably be a security
+_improvement_, since it would make it easier for the administrator to
+notice that it was not configured how they had wanted.
+
+## Design Details
+
+### kube-apiserver Command-Line Arguments
+
+```
+<<[UNRESOLVED apiserver-arguments ]>>
+
+Needs to be figured out, but probably:
+
+  - If `--bind-address` or `--advertise-address` is specified as a
+    single IP, then the apiserver is single-stack, even if dual
+    `--service-cluster-ip-range`s are specified.
+
+  - If `--bind-address` or `--advertise-address` is specified as a
+    pair of IPs, and `--service-cluster-ip-range` is specified as a
+    pair of CIDRs, then the apiserver is dual-stack, and should fail
+    if it cannot bind sockets for both address families.
+
+  - (Incompatible values for `--bind-address`, `--advertise-address`,
```

Comment:
I think when `--advertise-address` is dual-stack but `--bind-address` isn't then we **should** fail. Because simply you can not advertise address you are not bound to its family. But the other way around we *should not* fail. 

For example: A cluster may have monitoring tools running on control plane nodes that are running with alternative IP family. IOW I may want to bind to both family but i don't want to advertise both families.

Reply:
... unless you bind to 0.0.0.0 or :: , when you bind to all addresses and families, but you only one to advertise only specific ones